### PR TITLE
[DSW-161] Retry connect to MongoDB, RabbitMQ on start

### PR DESCRIPTION
**Pull request by MarekSuchanek**

### CODE REVIEW

**Comment by MarekSuchanek** [comment]
Now retry messages look like this:

```
...
[Info] SERVER: started
[Info] CONFIG: loaded
[Info] ENVIRONMENT: set to Production
[Info] DATABASE: creating connection pool
[Info] DATABASE: creating connection pool (retry #1)
[Info] DATABASE: creating connection pool (retry #2)
[Info] DATABASE: creating connection pool (retry #3)
[Info] DATABASE: creating connection pool (retry #4)
[Info] DATABASE: creating connection pool (retry #5)
```
and similarly
```
...
[Info] DATABASE: creating connection pool
[Info] DATABASE: connected
[Info] MESSAGING: connecting to messaging channel
[Info] MESSAGING: connecting to messaging channel (retry #1)
[Info] MESSAGING: connecting to messaging channel (retry #2)
[Info] MESSAGING: connecting to messaging channel (retry #3)
[Info] MESSAGING: connecting to messaging channel (retry #4)
...
```

Retry backoff is set to exponential (init with 1 second) and 5 retries (1, 2, 4, 8, and 16 seconds).

Is it suitable or something should be different?

**Comment by MarekSuchanek** [comment]
I've improved logging after our discussion...

```
[Info] SERVER: started
[Info] CONFIG: loaded
[Info] ENVIRONMENT: set to Production
[Info] DATABASE: creating a connection pool
[Info] DATABASE: connected
[Info] MESSAGING: connecting to the messaging channel
[Warn] MESSAGING: failed to connect to the messaging channel - retry #1 in 2.0 seconds
[Warn] MESSAGING: failed to connect to the messaging channel - retry #2 in 4.0 seconds
[Warn] MESSAGING: failed to connect to the messaging channel - retry #3 in 8.0 seconds
[Warn] MESSAGING: failed to connect to the messaging channel - retry #4 in 16.0 seconds
[Warn] MESSAGING: failed to connect to the messaging channel - retry #5 in 32.0 seconds
[Error] MESSAGING: failed to connect to the messaging channel
dsw-server: ConnectionClosedException Abnormal "Could not connect to any of the provided brokers: [((\"localhost\",5672),HostCannotConnect \"localhost\" [Network.Socket.connect: <socket: 3>: does not exist (Connection refused),Network.Socket.connect: <socket: 3>: does not exist (Connection refused)])]"
```

(similarly DB connection)

File: lib/Application.hs [hunk 1]
```diff
@@ -51,16 +58,48 @@ runServer =
...
-          lift . logInfo $ msg _CMP_DATABASE "connected"
-          msgChannel <- liftIO $ createMessagingChannel dswConfig
-          lift . logInfo $ msg _CMP_MESSAGING "connected"
-          let serverPort = dswConfig ^. webConfig ^. port
-          let baseContext =
-                BaseContext
-                {_baseContextConfig = dswConfig, _baseContextPool = dbPool, _baseContextMsgChannel = msgChannel}
-          liftIO $ runDBMigrations baseContext
-          liftIO $ runApplication baseContext
+        logInfo $ msg _CMP_DATABASE "creating a connection pool"
```
Comment (vknaisl) [inline lib/Application.hs]: I would change to `connecting to the database` when we have a `connecting to the message broker` message for the message broker

File: lib/Application.hs [hunk 2]
```diff
@@ -51,16 +58,48 @@ runServer =
...
+        dbPool <-
+          liftIO $
+          withRetry
+            retryBackoff
+            _CMP_DATABASE
+            "failed to create a connection pool"
```
Comment (vknaisl) [inline lib/Application.hs]: same as above - please change it to `failed to connect to the database`

**Review by vknaisl — CHANGES_REQUESTED** [review]

File: lib/Application.hs (hunk 1 above)
Comment (MarekSuchanek) [reply lib/Application.hs]: Done in https://github.com/ds-wizard/dsw-server/pull/31/commits/3fcdbc632da1440e7a67bc323a9d32612467bfe4

File: lib/Application.hs (hunk 2 above)
Comment (MarekSuchanek) [reply lib/Application.hs]: Done in https://github.com/ds-wizard/dsw-server/pull/31/commits/3fcdbc632da1440e7a67bc323a9d32612467bfe4

**Review by vknaisl — APPROVED** [review]

**Pull request merged**